File: frontend/whatsinit/lib/services/api_service.dart
import 'dart:convert';
import 'dart:io';
import 'dart:async';
import 'package:http/http.dart' as http;
import '../models/product.dart';
import '../models/analysis_result.dart';
import '../models/user_preferences.dart';

class ApiService {
  // In a real app, this would come from environment variables
  late String baseUrl;
  // Fixed backend URL
  static const String BACKEND_URL = 'https://api.whats-in-it.org';
  
  // Constructor that initializes the baseUrl
  ApiService() {
    baseUrl = BACKEND_URL;
  }
  
  
  
  
  // Get product information by barcode
  Future<Product> getProductByBarcode(String barcode) async {
    print('Attempting to connect to: $baseUrl/api/v1/product/$barcode');
    
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/api/v1/product/$barcode'),
        headers: {'Content-Type': 'application/json'},
      ).timeout(const Duration(seconds: 45)); // Increased timeout from 30 to 45 seconds

      print('Response received - Status: ${response.statusCode}');
      
      if (response.statusCode == 200) {
        return Product.fromJson(json.decode(response.body));
      } else {
        throw Exception('Failed to load product: ${response.statusCode} - ${response.body}');
      }
    } on SocketException catch (e) {
      print('Socket Exception: $e');
      throw Exception('Network error: Cannot connect to server. Please check your connection and try again.');
    } on TimeoutException catch (e) {
      print('Timeout Exception: $e');
      throw Exception('Connection timed out. The server is taking too long to respond.');
    } catch (e) {
      print('Generic Exception: $e');
      throw Exception('An unexpected error occurred: $e');
    }
  }

  // Get comprehensive analysis based on product and user preferences
  Future<AnalysisResult> getComprehensiveAnalysis(
      Product product, UserPreferences preferences) async {
    // Combine product data with user preferences for the request
    Map<String, dynamic> requestBody = {
      'product': {
        'barcode': product.barcode,
        'name': product.name,
        'brands': product.brand,
        'ingredients_text': product.ingredientsText,
        'ingredients_list': product.ingredientsList,
        'nutrition_facts': product.nutritionFacts,
      },
      'user_preferences': {
        'diet_type': preferences.dietType,
        'allergies': preferences.allergies,
        'avoid_ingredients': preferences.avoidIngredients,
        'health_concerns': {
          'sugar': preferences.sugarConcern,
          'salt': preferences.saltConcern,
          'fat': preferences.fatConcern,
        }
      }
    };

    // Log the request for debugging
    print('API Request to: $baseUrl/api/v1/analyze-comprehensive');
    print('Request body: ${json.encode(requestBody)}');

    try {
      final response = await http.post(
        Uri.parse('$baseUrl/api/v1/analyze-comprehensive'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode(requestBody),
      ).timeout(const Duration(seconds: 90)); // Increased timeout from 30 to 90 seconds for comprehensive analysis

      // Log the response for debugging
      print('Response status: ${response.statusCode}');
      print('Response body: ${response.body}');

      if (response.statusCode == 200) {
        return AnalysisResult.fromJson(json.decode(response.body));
      } else {
        throw Exception('Failed to get analysis: ${response.statusCode} - ${response.body}');
      }
    } catch (e) {
      print('Error during analysis request: $e');
      throw Exception('Failed to complete analysis: $e');
    }
  }
} 

File: frontend/web/src/services/apiService.ts
import { Product, AnalysisResult, UserPreferences, ComprehensiveAnalysisRequest, ApiError } from '@/types';

class ApiService {
  private readonly baseUrl = 'https://api.whats-in-it.org';
  
  private async request<T>(
    endpoint: string, 
    options: RequestInit = {}
  ): Promise<T> {
    const url = `${this.baseUrl}${endpoint}`;
    
    const config: RequestInit = {
      headers: {
        'Content-Type': 'application/json',
        ...options.headers,
      },
      ...options,
    };

    console.log(`API Request to: ${url}`);
    if (config.body) {
      console.log('Request body:', config.body);
    }

    try {
      const response = await fetch(url, config);
      
      console.log(`Response status: ${response.status}`);
      
      if (!response.ok) {
        const errorData = await response.text();
        console.log('Error response:', errorData);
        
        let errorMessage = `HTTP ${response.status}: ${response.statusText}`;
        
        try {
          const parsedError = JSON.parse(errorData);
          if (parsedError.detail) {
            if (Array.isArray(parsedError.detail)) {
              errorMessage = parsedError.detail.map((err: any) => err.msg).join(', ');
            } else if (typeof parsedError.detail === 'string') {
              errorMessage = parsedError.detail;
            }
          }
        } catch {
          // Use the raw error text if it's not JSON
          if (errorData) {
            errorMessage = errorData;
          }
        }

        const apiError: ApiError = {
          message: errorMessage,
          status: response.status,
        };
        
        throw apiError;
      }

      const data = await response.json();
      console.log('Response data:', data);
      
      return data;
    } catch (error) {
      console.error('API request failed:', error);
      
      if (error instanceof TypeError && error.message.includes('fetch')) {
        throw new Error('Network error: Cannot connect to server. Please check your connection and try again.');
      }
      
      // Re-throw ApiError as-is
      if (error && typeof error === 'object' && 'message' in error) {
        throw error;
      }
      
      throw new Error(`An unexpected error occurred: ${error}`);
    }
  }

  async getProductByBarcode(barcode: string): Promise<Product> {
    const timeoutController = new AbortController();
    const timeoutId = setTimeout(() => timeoutController.abort(), 45000); // 45 second timeout
    
    try {
      const product = await this.request<Product>(
        `/api/v1/product/${barcode}`,
        { signal: timeoutController.signal }
      );
      
      return product;
    } catch (error) {
      if (error instanceof DOMException && error.name === 'AbortError') {
        throw new Error('Connection timed out. The server is taking too long to respond.');
      }
      throw error;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  async getComprehensiveAnalysis(
    product: Product, 
    preferences: UserPreferences
  ): Promise<AnalysisResult> {
    const timeoutController = new AbortController();
    const timeoutId = setTimeout(() => timeoutController.abort(), 90000); // 90 second timeout for analysis
    
    try {
      const requestBody: ComprehensiveAnalysisRequest = {
        product: {
          barcode: product.barcode,
          name: product.name,
          brand: product.brand,
          image_url: product.image_url,
          ingredients_text: product.ingredients_text,
          ingredients_list: product.ingredients_list,
          nutrition_facts: product.nutrition_facts,
        },
        user_preferences: {
          diet_type: preferences.dietType,
          allergies: preferences.allergies,
          avoid_ingredients: preferences.avoidIngredients,
          health_concerns: {
            sugar: preferences.sugarConcern,
            salt: preferences.saltConcern,
            fat: preferences.fatConcern,
          },
        },
      };

      const analysis = await this.request<AnalysisResult>(
        '/api/v1/analyze-comprehensive',
        {
          method: 'POST',
          body: JSON.stringify(requestBody),
          signal: timeoutController.signal,
        }
      );

      return analysis;
    } catch (error) {
      if (error instanceof DOMException && error.name === 'AbortError') {
        throw new Error('Analysis timed out. The server is taking too long to complete the analysis.');
      }
      throw error;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  // Health check method
  async checkHealth(): Promise<{ status: string }> {
    try {
      return await this.request('/health');
    } catch (error) {
      throw new Error('API health check failed');
    }
  }

  // Format error messages for display
  static formatErrorMessage(error: unknown): string {
    if (typeof error === 'string') {
      return error;
    }
    
    if (error && typeof error === 'object' && 'message' in error) {
      const apiError = error as ApiError;
      const message = apiError.message;
      
      // Simplify common error messages
      if (message.includes('Failed to get analysis')) {
        if (message.includes('422')) {
          return 'The server could not process the request. Please check your product information and try again.';
        } else if (message.includes('404')) {
          return 'The analysis service could not be found. Please try again later.';
        } else if (message.includes('500')) {
          return 'The server encountered an error. Please try again later.';
        }
      }
      
      if (message.includes('Network error')) {
        return 'Cannot connect to the server. Please check your internet connection and try again.';
      }
      
      if (message.includes('timed out') || message.includes('timeout')) {
        return 'The request took too long to complete. Please try again.';
      }
      
      return message;
    }
    
    return 'An unexpected error occurred. Please try again.';
  }
}

// Create a singleton instance
export const apiService = new ApiService();
export default ApiService;